Marlin/src/module/tool_change.cpp

```diff
@@ -494,10 +494,15 @@ void tool_change(const uint8_t tmp_extruder, const float fr_mm_s/*=0.0*/, bool n
 
         // Move to the "old position" (move the extruder into place)
         if (!no_move && IsRunning()) {
+          #if ENABLED(SWITCHING_NOZZLE)
+            destination[Z_AXIS] += z_diff;  // Include the Z restore with the "move back"
+            z_diff = z_raise;               // and don't move again below.
+          #endif
           #if ENABLED(DEBUG_LEVELING_FEATURE)
             if (DEBUGGING(LEVELING)) DEBUG_POS("Move back", destination);
           #endif
-          prepare_move_to_destination();
+          // Move back to the original (or tweaked) position
+          do_blocking_move_to(destination[X_AXIS], destination[Y_AXIS], destination[Z_AXIS]);
```

Comment:
This will move Z first if it needs to raise. If it needs to lower, then XY will be done first. If it neither needs to raise or lower then only XY move is done.

Since this is the "move back" move (`destination` contains the original position), and no move will have been done up to now (except possibly a Z raise) then this should be safe in all cases.